### ISSUE TITLE
wp cv ext commands review and comments

### ISSUE THREAD

**Issue by kcristiano** [christianwach/cli-tools-for-civicrm#1]
tl;dr:   Really good implementation.  Commands work as documented.   Could use some improvements to ease adoption from those transitioning from `cv`

```
usage: 
wp cv ext disable <key-or-name> [--extpath=<extpath>] [--yes]
   or: wp cv ext download <key-or-name> [--install] [--yes]
   or: wp cv ext enable <key-or-name> [--extpath=<extpath>] [--yes]
   or: wp cv ext info <key-or-name> [--local] [--remote] [--format=<format>]
   or: wp cv ext install <key-or-name> [--extpath=<extpath>] [--yes]
   or: wp cv ext list [--local] [--remote] [--refresh] [--format=<format>]
   or: wp cv ext uninstall <key-or-name> [--extpath=<extpath>] [--yes]
   or: wp cv ext update-db
```

- `wp cv ext list`
  - Can we support `fields=`  such as `wp plugin list --fields=name,status,update,version` otherwise we receive a bit too much information (wraps on the screen)
  - typical usage in `cv` is `cv ext:list -L --columns=key,name,status,version` would be good to do `wp cv ext:list -L --fields=key,name,status,version`
  - `wp cv ext` is looking for `--local` can we add an alias for `--L`  as well as `--remote` as `--R`
- `wp cv ext download` `wp cv ext enable` `wp cv ext install` `wp cv ext uninstall` only supports a single extension at a time, `cv` allows multiple.
- `wp cv ext install` `wp cv ext enable`  These look to be aliases of one another.   Is that correct?
- `wp cv ext update-db --vvv`  is not verbose like `cv ext:upgrade-db -vvv`

**Comment by christianwach** [comment]
Thanks for reviewing @kcristiano - I appreciate it.

**Comment by christianwach** [comment]
@kcristiano Some progress for you to test...

> typical usage in `cv` is `cv ext:list -L --columns=key,name,status,version` would be good to do `wp cv ext:list -L --fields=key,name,status,version`

I've implemented a `--fields` argument so you can now do one of the following:

```
wp cv ext list --local --fields=key,name,status,version
wp cv ext list --local --columns=key,name,status,version
```

> `wp cv ext` is looking for `--local` can we add an alias for `--L` as well as `--remote` as `--R`

WP-CLI options must be lowercase, so `--L` and `--R` aren't possible. I've added lowercase `--l` and `--r` instead, e.g.

```
wp cv ext list --l --fields=key,name,status,version
wp cv ext list --l --columns=key,name,status,version
```

> `wp cv ext install` and `wp cv ext enable` These look to be aliases of one another. Is that correct?

Yes they are basically identical although they have different feedback text. They also make different API calls (`civicrm api extension.install` and `civicrm api extension.enable` respectively) so if these happen to diverge in future, this code will still be robust.

> `wp cv ext update-db --vvv` is not verbose like `cv ext:upgrade-db -vvv`

When I run `cv ext:upgrade-db`, the `-v` flag doesn't alter output at all. The `-vv` flag only tells me the PHP version and does something called "Checking Box requirements" which I'm unfamiliar with. I think these can be ignored.

I've implemented the `--vvv` argument so you can now do the following:

```
wp cv ext update-db --vvv
```

Have I missed anything?

**Comment by christianwach** [comment]
Regarding `civicrm api extension.install` and `civicrm api extension.enable` I would recommend using `install` as it appears to have [newer code](https://github.com/civicrm/civicrm-core/blob/538cabb8f3d1530680927e539cbee7d4217bfe5b/api/v3/Extension.php#L34) than `enable` which [doesn't wrap](https://github.com/civicrm/civicrm-core/blob/538cabb8f3d1530680927e539cbee7d4217bfe5b/api/v3/Extension.php#L112) the Extension system manager call in `try {} catch {}`.

**Comment by christianwach** [comment]
FWIW I've [opened a PR](https://github.com/civicrm/civicrm-core/pull/33084) to improve `civicrm api extension.enable`.

**Comment by christianwach** [comment]
Note that for `wp cv ext update-db` you have to specify a WordPress User or you'll hit [this issue](https://lab.civicrm.org/dev/core/-/issues/5713). Something like this works:

```
wp cv ext update-db --vvv  --user=1
```

Assuming `1` is your admin or super admin user. Change as needed. I'll add to the command docs.

**Comment by kcristiano** [comment]
Thanks @christianwach 

One question before I test this.  We can currently do `cv ext:download {extkey} --no-install`  to download or not install as well as `cv ext:download (extkey} --no-install --force` to force/overwrite an extension (upgrade).   

The `--no-install` is useful - is there an equivilant for the `wp cv ext` commands?

**Comment by christianwach** [comment]
> The `--no-install` is useful - is there an equivilant for the `wp cv ext` commands?

@kcristiano I actually found the default in `cv` to be at odds with the name of the command - it should more properly be called `cv ext:download-and-install` - and as a result it's just too easy to install an Extension you didn't mean to install. 

So `wp cv ext download` defaults to *just downloading the Extension and no more*. There is a flag, however, if you want to install as well - unsurprisingly called `--install` 😄  See `wp help cv ext download` for details.

**Comment by christianwach** [comment]
> `wp cv ext install` `wp cv ext enable` These look to be aliases of one another. Is that correct?

@kcristiano FYI [interesting comments on the PR](https://github.com/civicrm/civicrm-core/pull/33084) regarding this. The answer is "sort of but not really".

**Comment by kcristiano** [comment]
I did see that answer on the PR.   Well, it's what we have...

**Comment by kcristiano** [comment]
@christianwach do we need any core changes to prepare this for release?   While I am sure `wp cv ext` can be improved, I do think it's better if we release. That will allow anyone that uses the plugin to deploy and test.

**Comment by christianwach** [comment]
[Release 1.0.2](https://github.com/christianwach/cli-tools-for-civicrm/releases/tag/1.0.2) is out.

**Comment by christianwach** [comment]
@kcristiano Where are we at with this? Is there anything else that `wp civicrm ext` should do? Or any other glaring inconsistencies?

**Comment by kcristiano** [comment]
with cv 

```
ext:download   Download and enable an extension
      ext:enable     Enable an extension
      ext:disable    Disable an extension
      ext:list       List extensions
      ext:uninstall  Uninstall an extension and purge its data
      ext:upgrade-db Apply DB upgrades for any extensions (DEPRECATED).
```

wp-cli

```
usage: wp cv ext disable <key-or-name> [--extpath=<extpath>] [--yes]
   or: wp cv ext download <key-or-name> [--install] [--yes]
   or: wp cv ext enable <key-or-name> [--extpath=<extpath>] [--yes]
   or: wp cv ext info <key-or-name> [--local] [--remote] [--format=<format>]
   or: wp cv ext install <key-or-name> [--extpath=<extpath>] [--yes]
   or: wp cv ext list [--local] [--l] [--remote] [--r] [--refresh] [--fields=<fields>] [--columns=<columns>] [--format=<format>]
   or: wp cv ext uninstall <key-or-name> [--extpath=<extpath>] [--yes]
   or: wp cv ext update-db [--vvv]
```

main difference is that 

`cv ext:download/install` will accept an array of values.   So I can `wp cv api System.check name="checkExtensionsUpdates"  |grep -Po '(?<=key=)[^"]*' > ~/cvlog/extup.log `  and then `cv dl -f $(cat ~/cvlog/extup.log) ; cv ext:upgrade-db -vv && wp cv api System.check name="checkExtensionsUpdates"   |grep -Po '(?<=key=)[^"]*'; echo "done"`

In other words using wp-cli (and some grep) I can get a list of all extensions that need updating and feed that list to `cv dl -f` and update all extensions in one go.  `wp cv ext download/install`  only take one key or name

Other commands as noted above also only take one argument but this is my current use case that does not match.

There are other features here that are better - download does just that - download no install.  That is an improvement.   

So it's non-critical.  It'd be nice to have but the `wp cv ext` commands all do exactly what they say.
